Tests/PerformanceTest/XCTestCase+InstructionsCount.swift

```diff
@@ -0,0 +1,68 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import _InstructionCounter
+
+extension XCTestCase {
+  private var baselineURL: URL {
+    if let baselineFile = ProcessInfo.processInfo.environment["BASELINE_FILE"] {
+      return URL(fileURLWithPath: baselineFile)
+    } else {
+      return URL(fileURLWithPath: #file)
+        .deletingLastPathComponent()
+        .appendingPathComponent("baselines.json")
+    }
+  }
+
+  func measureInstructions(_ baselineName: StaticString = #function, block: () -> Void, file: StaticString = #file, line: UInt = #line) throws {
+    let startInstructions = getInstructionsExecuted()
+    block()
+    let endInstructions = getInstructionsExecuted()
+    let numberOfInstructions = endInstructions - startInstructions
+    let strippedBaselineName = "\(baselineName)".replacingOccurrences(of: "()", with: "")
+
+    // If the is no data, we just continue the test
+    guard let data = try? Data(contentsOf: baselineURL) else {
+      return
+    }
+
+    let jsonDecoder = JSONDecoder()
+    let baselineMap = try jsonDecoder.decode([String: UInt64].self, from: data)
+
+    guard let baseline = baselineMap[strippedBaselineName] else {
+      XCTFail(
+        """
+        Missing baseline for \(strippedBaselineName)
+        with number of instructions \(numberOfInstructions)
+        """,
+        file: file,
+        line: line
+      )
+      return
+    }
+
+    let lowerBaseline = UInt64(Double(baseline) * 0.995)
+    let upperBaseline = UInt64(Double(baseline) * 1.005)
+    let baselineRange = lowerBaseline...upperBaseline
```

Comment:
Looking at the failure from your second CI run, the first think that I did was to calculate the difference in percents. So, I’m thinking that maybe we should also phrase the error message in those terms.

```swift
let baseline = 2624475912
let numberOfInstructions = 2640510088

let relativeDifference = Double(numberOfInstructions)/Double(baseline) - 1
let allowedDeviation = 0.005

assert(
  (-allowedDeviation..<allowedDeviation).contains(relativeDifference),
  "Number of instructions (\(numberOfInstructions)) deviated from baseline by \(String(format: "%.4f", relativeDifference * 100))%. The maximum allowed deviation in either direction is \(allowedDeviation * 100)%"
)

```

Reply:
I slight changed the assert message. I also added the name of the baseline. Just so it's easier to find.

Reply:
OK, great 👍🏽